### PR TITLE
feat: check for any Node in toHaveTextContent

### LINKED ISSUE #306: toHaveTextContent should support any Node
- `@testing-library/jest-dom` version: 5.11.5
- `node` version: 12.18.0
- `yarn` version: 1.22.5
- `react-testing-library` version: 11.1.0

### Relevant code or config:

```js
    const resourceLabel = getByText('Resource:');
    const valueElement = ensureExists(resourceLabel.nextSibling);
    expect(valueElement).toHaveTextContent('libxul.so');
```

### What you did:

I'm in the process of installing and using jest-dom in our project. I installed the eslint plugin which converted this code:
```js
    const resourceLabel = getByText('Resource:');
    const valueElement = ensureExists(resourceLabel.nextSibling);
    expect(valueElement.textContent).toBe('libxul.so');
```

### What happened:
Here is the error I get:
```
    received value must be an HTMLElement or an SVGElement.
    Received has type:  object
    Received has value: libxul.so

       95 |     const resourceLabel = getByText('Resource:');
       96 |     const valueElement = ensureExists(resourceLabel.nextSibling);
    >  97 |     expect(valueElement).toHaveTextContent('libxul.so');

```

### Reproduction:
https://codesandbox.io/s/react-testing-library-demo-forked-0puq7

### Problem description:

The property `textContent` is present in all `Node`s, not only `HTMLElement`. Therefore it would be nice if this expectation worked on all `Node`s too, especially text nodes.

See the MDN page about `textContent`:
https://developer.mozilla.org/en-US/docs/Web/API/Node/textContent

### Suggested solution:
Is removing this line enough:
https://github.com/testing-library/jest-dom/blob/c11f9c53bbf94c1aa48b401f4f523500e904e128/src/to-have-text-content.js#L8

Or should we replace it with a check for `instanceof Node` ?

### PR DESCRIPTION
**Pull request by julienw**
Fixes #306

I based it off #307 but changed a lot of things, that's why I didn't keep the initial author information.

**What, Why**: Please have a look at #306
**How**: I added a `checkNode` function in `utils` so that it can be reused elsewhere. I followed the style of the other functions and also the comment in #307 about checking that we have a window before using it.

**Checklist**:

- [X] Documentation
- [X] Tests
- [ ] Updated Type Definitions <-- Not sure where to check them?
- [X] Ready to be merged

### CODE REVIEW

File: src/utils.js [hunk 1]
```diff
@@ -31,24 +31,45 @@ class HtmlElementTypeError extends Error {
...
+  }
+}
+
+class NodeTypeError extends GenericTypeError {
+  constructor(...args) {
+    super('be a Node', ...args)
+  }
+}
+
+function checkHasWindow(htmlElement, ErrorClass, ...args) {
```
Comment (gnapse) [inline src/utils.js:52]: I think you could've kept the signature of this one intact as we only call it internally in this module, only once, and always to check for the HTML element check, and not the Node check.

Not a big deal or a blocking change request. Can stay as is. Maybe I'm missing something.

**Review by gnapse — APPROVED** [review]

**Comment by gnapse** [comment]
> Updated Type Definitions <-- Not sure where to check them?

This will need to be handled separately in https://github.com/DefinitelyTyped/DefinitelyTyped/tree/master/types/testing-library__jest-dom

File: README.md [hunk 2]
```diff
@@ -748,7 +748,8 @@ The usual rules of css precedence apply.
 toHaveTextContent(text: string | RegExp, options?: {normalizeWhitespace: boolean})
 ```
 
-This allows you to check whether the given element has a text content or not.
+This allows you to check whether the given node has a text content or not. This
+supports elements, but also text nodes and fragments.
 
 When a `string` argument is passed through, it will perform a partial
 case-sensitive match to the element content.
```
Comment (gnapse) [inline README.md:755]:
```suggestion
case-sensitive match to the node content.
```

File: src/utils.js (hunk 1 above)
Comment (julienw) [reply src/utils.js:52]: See line 63, I actually call it for the node check ;-)

Comment (gnapse) [reply src/utils.js:52]: yes, I know.

My point was that that was the only call to it. But never mind, it's ok as is. It was a bit of a nit-pick, and maybe a wrong one at that anyway.

**Comment by julienw** [comment]
> This will need to be handled separately in https://github.com/DefinitelyTyped/DefinitelyTyped/tree/master/types/testing-library__jest-dom

I see, looking at this I believe there's no change needed. But I'm not using TS (we're using Flow in our project) so I may miss something.

For Flow, what we put in `expect` is typed as `any` so everything passes.

**Comment by julienw** [comment]
Thanks for the quick review and merge!

**Comment by gnapse** [comment]
Thank you too for your contribution!

**Comment by gnapse** [comment]
@all-contributors add @julienw for code, test

File: src/utils.js (hunk 1 above)
Comment (julienw) [reply src/utils.js:52]: ah yeah I see what you mean :-) yeah, maybe it's a bit over-designed :-)

**Comment by nickmccurdy** [comment]
The type declarations live in [`@types/testing-library__jest-dom`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/testing-library__jest-dom). However as far as I understand, the type of the expected value isn't actually checked, so I don't think you need to update them in this case.

**Comment by RandinAk** [comment]
> Fixes #306
> 
> 
> 
> I based it off #307 but changed a lot of things, that's why I didn't keep the initial author information.
> 
> 
> 
> 
> 
> 
> 
> **What, Why**: Please have a look at #306
> 
> **How**: I added a `checkNode` function in `utils` so that it can be reused elsewhere. I followed the style of the other functions and also the comment in #307 about checking that we have a window before using it.
> 
> 
> 
> **Checklist**:
> 
> 
> 
> 
> 
> 
> 
> 
> 
> 
> 
> 
> 
> 
> 
> - [X] Documentation
> 
> - [X] Tests
> 
> - [ ] Updated Type Definitions <-- Not sure where to check them?
> 
> - [X] Ready to be merged
> 
>       
> 
> 
> 
> 
> 
>

**Comment by gnapse** [comment]
@RandinAk what's the purpose of your last comment? It seems to be mostly a quote of this PR's description, with no extra info added.

**Comment by gnapse** [comment]
@all-contributors add @julienw for code, test

**Pull request merged**